dht.go

```diff
@@ -274,8 +277,82 @@ func (dht *IpfsDHT) putLocal(key string, rec *recpb.Record) error {
 // Update signals the routingTable to Update its last-seen status
 // on the given peer.
 func (dht *IpfsDHT) Update(ctx context.Context, p peer.ID) {
-	logger.Event(ctx, "updatePeer", p)
-	dht.routingTable.Update(p)
+	conns := dht.host.Network().ConnsToPeer(p)
+	switch len(conns) {
+	default:
+		logger.Warning("unclear if we should add peer with multiple open connections to us to our routing table")
+	case 0:
+		logger.Error("called update on a peer with no connection")
+	case 1:
+		dht.UpdateConn(ctx, conns[0])
+	}
+}
+
+func (dht *IpfsDHT) UpdateConn(ctx context.Context, c network.Conn) {
+	logger.Event(ctx, "updatePeer", c.RemotePeer())
+	if dht.shouldAddPeerToRoutingTable(c) {
+		dht.routingTable.Update(c.RemotePeer())
+	}
+}
+
+func (dht *IpfsDHT) shouldAddPeerToRoutingTable(c network.Conn) bool {
+	if isRelayAddr(c.RemoteMultiaddr()) {
+		return false
+	}
+	return c.Stat().Direction == network.DirOutbound ||
+		dht.hasSensibleAddressesForPeer(c)
+}
+
+func (dht *IpfsDHT) hasSensibleAddressesForPeer(c network.Conn) bool {
+	addrs := dht.host.Peerstore().Addrs(c.RemotePeer())
+	if len(addrs) == 0 {
+		return false
+	}
+
+	var hasPublicAddr bool
+	for _, a := range dht.host.Peerstore().Addrs(c.RemotePeer()) {
+		if isRelayAddr(a) {
+			return false
+		}
+
+		if manet.IsPublicAddr(a) {
+			hasPublicAddr = true
+		}
+	}
+
+	if dht.peerIsOnSameSubnet(c) && !hasPublicAddr {
+		// TODO: for now, we can't easily tell if the peer on our subnet
+		// is dialable or not, so don't discriminate.
+		// ! It may be okay to not add these to our routing table...
+		return true
+	}
+
+	if !hasPublicAddr {
+		return false
+	}
+
+	return false
+}
+
+// taken from go-libp2p/p2p/host/relay
+func isRelayAddr(a ma.Multiaddr) bool {
+	isRelay := false
+
+	ma.ForEach(a, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case circuit.P_CIRCUIT:
+			isRelay = true
+			return false
+		default:
+			return true
+		}
+	})
+
+	return isRelay
+}
+
+func (dht *IpfsDHT) peerIsOnSameSubnet(c network.Conn) bool {
+	return manet.IsPrivateAddr(c.RemoteMultiaddr())
```

Comment:
How is this checking if we're on the same subnet as the peer? How about a `manet.IsPrivateAddr(c.RemoteMultiaddr()) && c.RemoteMultiaddr() in addrs` check here?

Reply:
if we are connected to someone, and the connection is a private IP address, we must be on the same subnet as them, as private IPs are never routed outside of their subnet

Reply:
duh, of course